src/transformers/integrations/sdpa_attention.py

```diff
@@ -68,13 +68,22 @@ def sdpa_attention_forward(
     if attention_mask is not None and attention_mask.ndim == 4:
         attention_mask = attention_mask[:, :, :, : key.shape[-2]]
 
-    # We dispatch to SDPA's Flash Attention or Efficient kernels via this `is_causal` if statement instead of an inline conditional assignment
-    # in SDPA to support both torch.compile's dynamic shapes and full graph options. An inline conditional prevents dynamic shapes from compiling.
-    # Note that it is important to check first for the shape, otherwise compile will fail with `argument 'is_causal' must be bool, not SymBool`
-    if is_causal is None:
-        # The last condition is for encoder (decoder) models which specify this by passing their own `is_causal` flag
-        # This is mainly due to those models having mixed implementations for encoder, decoder, and encoder-decoder attns
-        is_causal = query.shape[2] > 1 and attention_mask is None and getattr(module, "is_causal", True)
+    # Kwarg takes precedence over the defined module's attribute
+    # - Allows dynamic switching, e.g. when model's switch based on the model input type (CLIP)
+    # - Defaults to "normal" behavior for all attention types (encoder, decoder, cross)
+    is_causal = is_causal if is_causal is not None else getattr(module, "is_causal", True)
+
+    # SDPA's Flash Attention (and cuDNN) kernels rely on the `is_causal` flag. However, there are certain conditions:
+    # - Not in decoding phase (== full attention)
+    # - Attention mask is not to be provided (even if it is a causal pattern)
+    # - Internally, we marked this as compatible with causal, i.e. it is a decoder attention type
```

Comment:
Actually, isn't the decoding case valid as well? Looking [here](https://github.com/huggingface/transformers/blob/main/src/transformers/masking_utils.py#L253-L254), it seems that using `is_causal = True/False` would be the same with a single-row matrix, i.e. we take the whole row anyway

Or the comment on the masking is wrong, and it's simply that we can skip the mask as well, but then we need full attention on the single query token (I think it's this one, do you mind updating comment on the mask?)

```suggestion
    # - Not in decoding phase (otherwise we want full attention on the single query token)
    # - Attention mask is not to be provided (even if it is a causal pattern)
    # - Internally, we marked this as compatible with causal, i.e. it is a decoder attention type
```